src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/azure/vm/AzureVMsProvider.java

```diff
@@ -0,0 +1,295 @@
+                                                                                                                                                           instance.getImage() +
+                                                                                                                                                           "'")));
+
+        // Get the options (Optional by design)
+        Optional<Options> options = Optional.ofNullable(instance.getOptions());
+
+        // Try to retrieve the resourceGroup from provided name, otherwise get it from image
+        ResourceGroup resourceGroup = azureProviderUtils.searchResourceGroupByName(azureService,
+                                                                                   options.map(Options::getResourceGroup)
+                                                                                          .orElseGet(image::resourceGroupName))
+                                                        .orElseThrow(() -> new RuntimeException("ERROR unable to find a suitable resourceGroup from instance: '" +
+                                                                                                instance + "'"));
+
+        // Try to get region from provided name, otherwise get it from image
+        Region region = options.map(presentOptions -> Region.findByLabelOrName(presentOptions.getRegion()))
+                               .orElseGet(image::region);
+
+        // Prepare the VM(s)
+        Optional<Boolean> optionalStaticPublicIP = options.map(Options::getStaticPublicIP);
+        List<Creatable<VirtualMachine>> creatableVirtualMachines = prepareCreatableVirtualMachines(azureService,
+                                                                                                   options,
+                                                                                                   region,
+                                                                                                   resourceGroup,
+                                                                                                   instanceTag,
+                                                                                                   instance,
+                                                                                                   image,
+                                                                                                   optionalStaticPublicIP);
+
+        // Create all VMs in parallel and collect IDs
+        return azureService.virtualMachines()
+                           .create(creatableVirtualMachines)
+                           .values()
+                           .stream()
+                           .map(vm -> instance.withTag(vm.name()).withId(vm.vmId()).withNumber(SINGLE_INSTANCE_NUMBER))
+                           .collect(Collectors.toSet());
+    }
+
+    private List<Creatable<VirtualMachine>> prepareCreatableVirtualMachines(Azure azureService,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Method has 8 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00107)